node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,199 @@ async fn fetch_block_worker(
+fn are_signatures_sufficient_for_sync_to_genesis(
+    consensus_verdict: Result<(), FinalitySignatureError>,
+) -> bool {
+    match consensus_verdict {
+        Ok(_) | Err(FinalitySignatureError::TooManySignatures { .. }) => true,
+        Err(_) => false,
+    }
+}
+
+// Fetches the finality signatures from the given peer. In case of timeout, it'll
+// retry up to `retries` times. Other errors interrupt the process immediately.
+async fn fetch_finality_signatures_with_retry(
+    block_hash: BlockHash,
+    peer: NodeId,
+    retries: usize,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<FetchedData<BlockSignatures>, FetcherError<BlockSignatures>> {
+    for _ in 0..retries {
+        let maybe_signatures = ctx
+            .effect_builder
+            .fetch::<BlockSignatures>(block_hash, peer)
+            .await;
+        match maybe_signatures {
+            Ok(result) => return Ok(result),
+            Err(FetcherError::TimedOut { .. }) => continue,
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_header(
+    block_header: BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<Option<BlockSignatures>, Error> {
+    let start = Timestamp::now();
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
+
+    let mut sig_collector = BlockSignaturesCollector::new();
+
+    let block_header_hash = block_header.hash(ctx.config.verifiable_chunked_hash_activation());
+
+    for peer in peer_list {
+        let fetched_signatures = fetch_finality_signatures_with_retry(
+            block_header_hash,
+            peer,
+            FINALITY_SIGNATURE_FETCH_RETRY_COUNT,
+            ctx,
+        )
+        .await;
+        let maybe_signatures = match fetched_signatures {
+            Ok(FetchedData::FromStorage { item, .. }) => {
+                trace!(
+                    ?block_header_hash,
+                    ?peer,
+                    "did not get FinalitySignatures from peer, got from storage instead",
+                );
+                ctx.metrics
+                    .observe_fetch_finality_signatures_duration_seconds(start);
+                return Ok(Some(*item));
+            }
+            Ok(FetchedData::FromPeer { item, .. }) => {
+                trace!(?block_header_hash, ?peer, "fetched FinalitySignatures");
+                Some(*item)
+            }
+            Err(_) => None,
+        };
+
+        let era_for_validators_retrieval = get_era_id_for_validators_retrieval(
+            &block_header.era_id(),
+            ctx.config.last_emergency_restart(),
+        );
+
+        let maybe_switch_block_of_previous_era = ctx
+            .effect_builder
+            .get_switch_block_header_at_era_id_from_storage(era_for_validators_retrieval)
```

Comment:
Since we can get `None` here, which would already be an error, I would add another error variant and return it here if we don't get the expected switch block from storage. Makes for simpler debugging.

Reply:
Dedicated error variant added in https://github.com/casper-network/casper-node/pull/3125/commits/44dd95ceba0010eade9c1a4673ff5fff70c058dc